Cabal/src/Distribution/PackageDescription/Check/Prim.hs

```diff
@@ -0,0 +1,1164 @@
+import Distribution.Types.PackageId (PackageIdentifier)
+import Distribution.Types.PackageDescription (package, specVersion)
+import Distribution.Types.PackageName (PackageName)
+import Distribution.Types.LibraryName (LibraryName, showLibraryName)
+import Distribution.Types.TestType (TestType, knownTestTypes)
+import Distribution.Types.UnqualComponentName
+import Distribution.Types.Version (Version)
+import Distribution.Utils.Path (LicenseFile, PackageDir, SymbolicPath,
+                                getSymbolicPath)
+import Language.Haskell.Extension (Extension)
+
+import qualified Control.Monad.Reader as Reader
+import qualified Control.Monad.Writer as Writer
+import qualified Control.Monad.Trans.Class as Trans
+import qualified Data.ByteString.Lazy as BS
+import qualified Data.List as List
+import qualified Data.Set as Set
+
+-- ------------------------------------------------------------
+-- * Check monad
+-- ------------------------------------------------------------
+
+-- Monadic interface for for Distribution.PackageDescription.Check.
+--
+-- Monadic checking allows us to have a fine grained control on checks
+-- (e.g. omitting warning checks in certain situations).
+
+-- * Interfaces
+--
+
+-- | Which interface to we have available/should we use? (to perform: pure
+-- checks, package checks, pre-distribution checks.)
+data CheckInterface m =
+        CheckInterface { ciPureChecks :: Bool,
+                            -- Perform pure checks?
+                         ciPackageOps :: Maybe (CheckPackageContentOps m),
+                            -- If you want to perform package contents
+                            -- checks, provide an interface.
+                         ciPreDistOps :: Maybe (CheckPreDistributionOps m)
+                            -- If you want to work-tree checks, provide
+                            -- an interface.
+                         }
+
+-- | A record of operations needed to check the contents of packages.
+-- Abstracted over `m` to provide flexibility (could be IO, a .tar.gz
+-- file, etc).
+--
+data CheckPackageContentOps m = CheckPackageContentOps {
+    doesFileExist        :: FilePath -> m Bool,
+    doesDirectoryExist   :: FilePath -> m Bool,
+    getDirectoryContents :: FilePath -> m [FilePath],
+    getFileContents      :: FilePath -> m BS.ByteString
+  }
+
+-- | A record of operations needed to check contents *of the work tree*
+-- (compare it with 'CheckPackageContentOps'). This is still `m` abstracted
+-- in case in the future we can obtain the same infos other than from IO
+-- (e.g. a VCS work tree).
+--
+data CheckPreDistributionOps m = CheckPreDistributionOps {
+        runDirFileGlobM :: FilePath -> Glob -> m [GlobResult FilePath],
+        getDirectoryContentsM :: FilePath -> m [FilePath]
+    }
+
+-- | Context to perform checks (will be the Reader part in your monad).
+--
+data CheckCtx m = CheckCtx { ccInterface :: CheckInterface m,
```

Comment:
You suddenly change indentation style here: you leave the first field on the same line and then you have to bump all the rest much further to the right than it makes sense to me. Could you do the same as in other definitions, where the first field goes on the next line and the whole alignment doesn't go far right? (There's a space for a joke about far right here, perhaps.)